### PR TITLE
Update temp sensor code

### PR DESCRIPTION
**Pull request by khanr107**
temp data logging code

### CODE REVIEW

**Comment by ShyavanS** [comment]
Looks good, but where did you get the info for the kalman filter setup you are using? Mind sending over the source, just curious? Will any of the initial parameters you set need to be tuned as we do trial runs? Also make sure whoever is doing the competition codebase version of the temp sensor task & the IMU tasks also is on the same page on what you did for the kalman filtering. We ideally want to keep everything to the same algorithm to keep our filtering consistent across sensors & codebases.

Only issue I see with the code though is you'll have to update the code with the new pin assignments & replace 23 with A1. Don't forget when you do this to pull changes from the main branch to your branch too so they're both up-to-date as I made some changes to the main branch for the pcb.

File: Car_Code/cheme_car_datalogging_codebase/cheme_car_datalogging_codebase.ino [hunk 1]
```diff
@@ -1,15 +1,63 @@
...
+// Rida -In Progress- Temperature Sensor - Needs to be initialized, periodically poll sensor, store temperature in a variable, print out to serial monitor, look into Kalman filters for reducing sensor noise.
 // IMU - Needs to be initialized, periodically poll sensor, store Z-axis angle in variable, take initial value as 0 deg print out to serial monitor, look into kalman filters for reducing sensor noise.
 // Stir Bar Motor - Needs to be initialized, just constantly turn at 80% speed for now, we'll tune speed later.
 // Servo Motor - Needs to be initialized, turn 180 deg clockwise once at start to dump reactants, wait 1 sec to finish dumping, turn back 180 deg counter-clockwise to return to upright position, set speed to 100% for now, tune later.
 // MicroSD Card - Store all variable data to MicroSD Card on microcontroller, allow for each run to be saved under a different filename (potentially in CSV format) so we can keep track of everything.
 
+#include <OneWire.h>
+#include <DallasTemperature.h>
+#define ONE_WIRE_BUS 23 // pin for the DS18B20 data line
```
Comment (ShyavanS) [inline Car_Code/cheme_car_datalogging_codebase/cheme_car_datalogging_codebase.ino]: Change this from 23 to the new pin assignments (should be A1). Don't forget when you do this to pull changes from the main branch to your branch too so they're both up-to-date as I made some changes to the main branch for the pcb.

File: Car_Code/cheme_car_datalogging_codebase/cheme_car_datalogging_codebase.ino [hunk 2]
```diff
@@ -1,15 +1,63 @@
...
+
+  // Kalman filter update
+
+  /* Kalman gain: calculated based on the predicted error covariance 
+  and the measurement noise covariance,,, used to update the 
+  state estimate (x_k) and error covariance (p_k) */ 
+  float k = p_k_minus / (p_k_minus + r); // kalman gain
+  //comparison with actual temp reading 
+  x_k = x_k_minus + k * (temperatureC - x_k_minus); // Updated state estimate
+  p_k = (1 - k) * p_k_minus; // Updated error covariance
```
Comment (ShyavanS) [inline Car_Code/cheme_car_datalogging_codebase/cheme_car_datalogging_codebase.ino:52]: Where did you get the info for the kalman filter setup you are using? Mind sending over the source, just curious?

File: Car_Code/cheme_car_datalogging_codebase/cheme_car_datalogging_codebase.ino [hunk 3]
```diff
@@ -1,15 +1,63 @@
...
+
 void setup() {
+  Serial.begin(9600);  // start serial communication (adjust baud rate as needed)
+  sensors.begin();   // initialize the DS18B20 sensor
 
+  // Initialize Kalman filter parameters
+  x_k = 0.0; // Initial state estimate
+  p_k = 1.0; // Initial error covariance
+  q = 0.01;  // Process noise covariance 
+  r = 0.1;   // Measurement noise covariance
```
Comment (ShyavanS) [inline Car_Code/cheme_car_datalogging_codebase/cheme_car_datalogging_codebase.ino:32]: Will any of the initial parameters you set need to be tuned as we do trial runs?

**Review by ShyavanS — CHANGES_REQUESTED** [review]
Looks good for the most part, just have some questions & one thing to resolve. Also, make sure whoever is doing the competition codebase version of the temp sensor task & the IMU tasks also is on the same page on what you did for the kalman filtering. We ideally want to keep everything to the same algorithm to keep our filtering consistent across sensors & codebases.

File: Car_Code/cheme_car_datalogging_codebase/cheme_car_datalogging_codebase.ino (hunk 2 above)
Comment (khanr107) [reply Car_Code/cheme_car_datalogging_codebase/cheme_car_datalogging_codebase.ino:52]: I used this video  (pseudo code at 6:48)
https://www.youtube.com/watch?v=ruB917YmtgE&t=95s&ab_channel=VDEngineering

Comment (ShyavanS) [reply Car_Code/cheme_car_datalogging_codebase/cheme_car_datalogging_codebase.ino:52]: Okay sounds good!

File: Car_Code/cheme_car_datalogging_codebase/cheme_car_datalogging_codebase.ino (hunk 3 above)
Comment (khanr107) [reply Car_Code/cheme_car_datalogging_codebase/cheme_car_datalogging_codebase.ino:32]: Yes, these parameters need to be tuned based on the characteristics of the system and the noise in our measurements. All of these would be determined while doing trial run. like for example q needs to be increased in the case of our system being very dynamic.

**Review by ShyavanS — APPROVED** [review]
Looks good now! Should be good to merge once Preet reads over & approves it.

**Review by batrap5 — APPROVED** [review]

**Pull request merged**